lib/matplotlib/legend.py

```diff
@@ -480,6 +482,11 @@ def val_or_rc(val, rc_name):
         self._mode = mode
         self.set_bbox_to_anchor(bbox_to_anchor, bbox_transform)
 
+        # Figure out if self.shadow is valid
+
+        if not (is_color_like(self.shadow) or isinstance(self.shadow, bool)):
```

Comment:
I'm a little wary of going to strict type checking, there is a strong expectations that `0` and `1` will work basically anyplace `True` / `False` will.

Reply:
I am under the impression that `0` and `1` are all one might expect to work, as opposed to options like `"true"`, `"false"` as in `validate.bool` in [`rcsetup.py`](https://github.com/matplotlib/matplotlib/blob/v3.6.2/lib/matplotlib/rcsetup.py#L140-L149).
Do you think it is sufficient to change `isinstance` to `shadow in (0, 1, True, False)`?

Reply:
@tacaswell I'm somewhat annoyed by the `0`, `1` topic.

In many cases (also here before the PR), we do implicit bool-casting via `if shadow:`. That means users can pass `0` or `[]` or whatever they like. I don't think we give a guarantee that these will work in the future.

In my personal opinion using `0`, `1` as a bool substitute is an anti-pattern that reduces readability. (e.g. here it could also be the size of the shadow).

Anyway, what I definitively don't want is cluttering our code internally with `0`, `1` checks.

We should

- either stay as permissive as before (i.e. do not do any check here), which is somewhat problematic as 
- or do the strict type checking

If you insist that `0`, `1` is really worth supporting, we should have a general discussion on that in the dev call. We'd need to document what exactly we want to accept in such cases in the dev guidelines. And we should come up with supporting functionality like `is_bool_like()`.

Reply:
I would suggest the correct way to check for T/F is perhaps something like `np.ndim(x) == 0 and (self.shadow == False or self.shadow == True)`?  (where the np.ndim check is only to avoid problems with array-equality returning bool arrays that cannot be used in a bool scalar context).  This way you take advantage of the fact that True == 1 and False == 0 (a fact that also makes supporting T/F but not 0/1 a bit weird; note that the same argument does not apply to other truthy or falsy values).